Port documentation from Wiki: Content query syntax (SenseNet#687) (#690)
SenseNet/sensenet: docs/content-query-syntax.md

---

> title:  "Content Query Syntax "

Reviewer: ```suggestion
title:  "Content Query Syntax"
```

source_url: 'https://github.com/SenseNet/sensenet/blob/master/docs/content-query-syntax.md'
category: Concepts
version: v6.0
tags: [content, query, syntax]
---

# Content Query Syntax 

## Overview

> Sense/Net uses the *Lucene search engine* for indexing and searching for content. We fully support the original [Lucene syntax](http://lucene.apache.org/java/2_9_1/queryparsersyntax.html), but extended it with our own keywords for sorting and filtering, thus forming the content query language (CQL). For the overview of the concept of querying in Sense/Net, visit the [Content Query](https://github.com/SenseNet/sensenet/blob/master/docs/field-setting.md) page.

Reviewer: ```suggestion
sensenet uses the *Lucene search engine* for indexing and searching for content. We fully support the original [Lucene syntax](http://lucene.apache.org/java/2_9_1/queryparsersyntax.html), but extended it with our own keywords for sorting and filtering, thus forming the content query language (CQL). For the overview of the concept of querying in sensenet, visit the [Content Query](https://github.com/SenseNet/sensenet/blob/master/docs/field-setting.md) page.
```

## Proximity search
It is possible to limit the distance of the relevant words in the result documents. The proximity search is two word in quots and followed by "~" and a positive whole number. This number defines the maximal count of words between the given two word. Example:
```bash
Description:"Lorem amet"~3
```
This query will find the following text: "Lorem ipsum dolor sit amet" because there is 3 words between "Lorem" and "amet".

## Fields
When performing a search you can specify the field you want to search in. In the previous cases Lucene used the **default field** called *_Text*, which contains all relevant data of a content. If you search for apple, it is the same as *_Text:apple*.

You can name any existing field, the syntax is *FieldName:VALUE*.
```bash
DisplayName:about
```
There are a few special field types that have their own syntax for defining the query term. For example **DateTime**:
```bash
ModificationDate:>'2010-09-01 12:00:00'
```
If you want to know more about how fields are indexed, visit the [Field Indexing](https://github.com/SenseNet/sensenet/blob/master/docs/field-indexing.md)  article.

### Common search-related fields

> Sense/Net has lots of built-in fields that you can use to make more precise queries. These fields are related to type or place of the content:

Reviewer: ```suggestion
sensenet has lots of built-in fields that you can use to make more precise queries. These fields are related to type or place of the content:
```

| **Name**		| **Comment**													| **Example**						|
| -------------	| ------------------------------------------------------------- | ---------------------------------	|
| **InTree**	| Returns the given folder and all contents under it			| InTree:"/Root/MyContent"			|
| **InFolder**	| Returns all content in that particular folder					| InFolder:"/Root/MyContent/Media"	|
| **Type**		| Returns1 all content that is of the given type				| Type:Article						|
| **TypeIs**	| Returns all content that is of the given type or any subtype	| TypeIs:WebContent					|

> > It is usually advised to use quotation marks when querying paths, since paths containing spaces and other special characters like '-' or '_' could cause trouble and "split up" the expression into two separate expressions instead of handling it as one query for path.}}

Reviewer: ```suggestion
> It is usually advised to use quotation marks when querying paths, since paths containing spaces and other special characters like '-' or '\_' could cause trouble and "split up" the expression into two separate expressions instead of handling it as one query for path.}}
```

```bash
Ingredients:apple AND Ingredients:melon
```
**+**

The "**+**" or **required** operator requires that the term after the "+" symbol exist somewhere in a the field of a single document.

To search for content that must contain "apple" and may contain "melon" use the query:
```bash
+apple melon
```
**NOT**

The **NOT** operator excludes content that contain the term after NOT. It cannot be used with only one term.
```bash
apple NOT melon
```
**-**

The "**-**" or **prohibit** operator excludes documents that contain the term after the "-" symbol.

To search for content that contain "apple" but not "melon" use the query:
```bash
apple -melon
```
## Grouping
Lucene supports using parentheses to group clauses to form sub queries. This can be very useful if you want to control the boolean logic for a query.
```bash
(lime AND apple) OR melon
```
## Paging
One of the most important things when writing queries is paging. In most cases you will not need all of the content but only a few of them that you can present on the UI.

If you want to display only the first 5 content, you can use the keyword **TOP**:
```bash
InFolder:/Root/MyContents .TOP:5
```

> > Always use the TOP keyword whenever limiting result count is possible. TOP limits the size of temporary result arrays used in background query execution logic. Not providing .TOP keyword may result in unnecessary memory consumption and unreasonably high intensity GC usage! See Query Optimization for details.}}

Reviewer: ```suggestion
> Always use the TOP keyword whenever limiting result count is possible. TOP limits the size of temporary result arrays used in background query execution logic. Not providing .TOP keyword may result in unnecessary memory consumption and unreasonably high intensity GC usage! See Query Optimization for details.
```

In case you want to display a few content but not the first ones (for example in a box showing related articles or you create a user interface that lets the user choose a page), use the keyword **SKIP**. The query below will skip the first 3 results and will return the second 3:
```bash
InFolder:/Root/MyContents .SKIP:3 .TOP:3
```
## Range search
Range queries allow one to match documents whose field(s) values are between the lower and upper bound specified by the range query. Range queries can be inclusive or exclusive of the upper and lower bounds. As you can see in the examples below, the two types can be mixed.

-   **[...TO...]**: inclusive (the lower and upper bounds are included)
-   **{...TO...}**: exclusive (the lower and upper bounds are NOT included)
```bash
CreationDate:['2010-08-30' TO '2010-10-30']
CreationDate:{'2010-08-30' TO '2010-10-30'}
CreationDate:['2010-08-30' TO '2010-10-30'}
CreationDate:{'2010-08-30' TO '2010-10-30']
CreationDate:>'2010-08-30'
CreationDate:<='2010-10-30'
```
## Filter keywords

> Sense/Net has several types of filters that automatically filter the results. These filters are **switched ON** by default, but you can decide to switch them off.

Reviewer: ```suggestion
sensenet has several types of filters that automatically filter the results. These filters are **switched ON** by default, but you can decide to switch them off.
```
